File: src/main/java/com/uiptv/service/remotesync/SecureTempFileSupport.java
package com.uiptv.service.remotesync;

import java.io.File;
import java.io.IOException;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.DosFileAttributeView;
import java.nio.file.attribute.FileAttribute;
import java.nio.file.attribute.PosixFilePermission;
import java.nio.file.attribute.PosixFilePermissions;
import java.util.Set;

final class SecureTempFileSupport {
    private static final String READ_PERMISSION = "read";
    private static final String WRITE_PERMISSION = "write";
    private static final String EXECUTE_PERMISSION = "execute";
    private static final Set<PosixFilePermission> OWNER_ONLY_DIRECTORY_PERMISSIONS = Set.of(
            PosixFilePermission.OWNER_READ,
            PosixFilePermission.OWNER_WRITE,
            PosixFilePermission.OWNER_EXECUTE
    );
    private static final Set<PosixFilePermission> OWNER_ONLY_FILE_PERMISSIONS = Set.of(
            PosixFilePermission.OWNER_READ,
            PosixFilePermission.OWNER_WRITE
    );
    private static final Path APP_TEMP_DIRECTORY = Path.of(
            System.getProperty("user.home", "."),
            ".uiptv",
            "tmp"
    );

    private SecureTempFileSupport() {
    }

    static Path createTempFile(String prefix, String suffix) throws IOException {
        Path tempDirectory = ensureAppTempDirectory();
        try {
            FileAttribute<Set<PosixFilePermission>> permissions =
                    PosixFilePermissions.asFileAttribute(OWNER_ONLY_FILE_PERMISSIONS);
            return Files.createTempFile(tempDirectory, prefix, suffix, permissions);
        } catch (UnsupportedOperationException _) {
            Path path = Files.createTempFile(tempDirectory, prefix, suffix);
            ensureOwnerOnlyFileAccess(path);
            return path;
        }
    }

    private static Path ensureAppTempDirectory() throws IOException {
        try {
            FileAttribute<Set<PosixFilePermission>> permissions =
                    PosixFilePermissions.asFileAttribute(OWNER_ONLY_DIRECTORY_PERMISSIONS);
            return Files.createDirectories(APP_TEMP_DIRECTORY, permissions);
        } catch (UnsupportedOperationException _) {
            Files.createDirectories(APP_TEMP_DIRECTORY);
            ensureOwnerOnlyDirectoryAccess(APP_TEMP_DIRECTORY);
            return APP_TEMP_DIRECTORY;
        } catch (FileAlreadyExistsException _) {
            ensureOwnerOnlyDirectoryAccess(APP_TEMP_DIRECTORY);
            return APP_TEMP_DIRECTORY;
        }
    }

    private static void ensureOwnerOnlyDirectoryAccess(Path path) throws IOException {
        try {
            Files.setPosixFilePermissions(path, OWNER_ONLY_DIRECTORY_PERMISSIONS);
        } catch (UnsupportedOperationException _) {
            ensureBestEffortOwnerOnlyDirectoryAccess(path);
        }
    }

    private static void ensureOwnerOnlyFileAccess(Path path) throws IOException {
        try {
            Files.setPosixFilePermissions(path, OWNER_ONLY_FILE_PERMISSIONS);
        } catch (UnsupportedOperationException _) {
            ensureBestEffortOwnerOnlyFileAccess(path);
        }
    }

    private static void ensureBestEffortOwnerOnlyDirectoryAccess(Path path) throws IOException {
        File file = path.toFile();
        applyBestEffortOwnerOnlyAccess(file, true);
        if (!Files.isDirectory(path)) {
            throw new IOException("Unable to prepare temp directory " + path);
        }
    }

    private static void ensureBestEffortOwnerOnlyFileAccess(Path path) throws IOException {
        File file = path.toFile();
        applyBestEffortOwnerOnlyAccess(file, false);
        if (!Files.isRegularFile(path)) {
            throw new IOException("Unable to prepare temp file " + path);
        }
    }

    private static void applyBestEffortOwnerOnlyAccess(File file, boolean executable) throws IOException {
        file.setReadable(false, false);
        file.setWritable(false, false);
        file.setExecutable(false, false);

        boolean ownerReadable = file.setReadable(true, true);
        boolean ownerWritable = file.setWritable(true, true);
        boolean ownerExecutable = !executable || file.setExecutable(true, true);

        if (!ownerReadable || !ownerWritable || !ownerExecutable) {
            DosFileAttributeView dosView = Files.getFileAttributeView(file.toPath(), DosFileAttributeView.class);
            if (dosView != null) {
                return;
            }

            String permission = !ownerReadable ? READ_PERMISSION : !ownerWritable ? WRITE_PERMISSION : EXECUTE_PERMISSION;
            throw new IOException("Unable to set " + permission + " permissions for " + file.toPath());
        }
    }
}

File: server/src/main/kotlin/com/uiptv/service/remotesync/SecureTempFileSupport.kt
package com.uiptv.service.remotesync

import java.io.File
import java.io.IOException
import java.nio.file.FileAlreadyExistsException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.attribute.DosFileAttributeView
import java.nio.file.attribute.FileAttribute
import java.nio.file.attribute.PosixFilePermission
import java.nio.file.attribute.PosixFilePermissions

internal object SecureTempFileSupport {
    private const val READ_PERMISSION = "read"
    private const val WRITE_PERMISSION = "write"
    private const val EXECUTE_PERMISSION = "execute"
    private val ownerOnlyDirectoryPermissions = setOf(
        PosixFilePermission.OWNER_READ,
        PosixFilePermission.OWNER_WRITE,
        PosixFilePermission.OWNER_EXECUTE
    )
    private val ownerOnlyFilePermissions = setOf(
        PosixFilePermission.OWNER_READ,
        PosixFilePermission.OWNER_WRITE
    )
    private val appTempDirectory: Path = Path.of(System.getProperty("user.home", "."), ".uiptv", "tmp")

    @JvmStatic
    @Throws(IOException::class)
    fun createTempFile(prefix: String, suffix: String): Path {
        val tempDirectory = ensureAppTempDirectory()
        return try {
            val permissions: FileAttribute<Set<PosixFilePermission>> =
                PosixFilePermissions.asFileAttribute(ownerOnlyFilePermissions)
            Files.createTempFile(tempDirectory, prefix, suffix, permissions)
        } catch (_: UnsupportedOperationException) {
            val path = Files.createTempFile(tempDirectory, prefix, suffix)
            ensureOwnerOnlyFileAccess(path)
            path
        }
    }

    @Throws(IOException::class)
    private fun ensureAppTempDirectory(): Path {
        return try {
            val permissions: FileAttribute<Set<PosixFilePermission>> =
                PosixFilePermissions.asFileAttribute(ownerOnlyDirectoryPermissions)
            Files.createDirectories(appTempDirectory, permissions)
        } catch (_: UnsupportedOperationException) {
            Files.createDirectories(appTempDirectory)
            ensureOwnerOnlyDirectoryAccess(appTempDirectory)
            appTempDirectory
        } catch (_: FileAlreadyExistsException) {
            ensureOwnerOnlyDirectoryAccess(appTempDirectory)
            appTempDirectory
        }
    }

    @Throws(IOException::class)
    private fun ensureOwnerOnlyDirectoryAccess(path: Path) {
        try {
            Files.setPosixFilePermissions(path, ownerOnlyDirectoryPermissions)
        } catch (_: UnsupportedOperationException) {
            ensureBestEffortOwnerOnlyDirectoryAccess(path)
        }
    }

    @Throws(IOException::class)
    private fun ensureOwnerOnlyFileAccess(path: Path) {
        try {
            Files.setPosixFilePermissions(path, ownerOnlyFilePermissions)
        } catch (_: UnsupportedOperationException) {
            ensureBestEffortOwnerOnlyFileAccess(path)
        }
    }

    @Throws(IOException::class)
    private fun ensureBestEffortOwnerOnlyDirectoryAccess(path: Path) {
        val file = path.toFile()
        applyBestEffortOwnerOnlyAccess(file, true)
        if (!Files.isDirectory(path)) {
            throw IOException("Unable to prepare temp directory $path")
        }
    }

    @Throws(IOException::class)
    private fun ensureBestEffortOwnerOnlyFileAccess(path: Path) {
        val file = path.toFile()
        applyBestEffortOwnerOnlyAccess(file, false)
        if (!Files.isRegularFile(path)) {
            throw IOException("Unable to prepare temp file $path")
        }
    }

    @Throws(IOException::class)
    private fun applyBestEffortOwnerOnlyAccess(file: File, executable: Boolean) {
        file.setReadable(false, false)
        file.setWritable(false, false)
        file.setExecutable(false, false)

        val ownerReadable = file.setReadable(true, true)
        val ownerWritable = file.setWritable(true, true)
        val ownerExecutable = !executable || file.setExecutable(true, true)

        if (!ownerReadable || !ownerWritable || !ownerExecutable) {
            val dosView = Files.getFileAttributeView(file.toPath(), DosFileAttributeView::class.java)
            if (dosView != null) {
                return
            }
            val permission = if (!ownerReadable) READ_PERMISSION else if (!ownerWritable) WRITE_PERMISSION else EXECUTE_PERMISSION
            throw IOException("Unable to set $permission permissions for " + file.toPath())
        }
    }
}
